File: public/javascripts/component/data/repeater.js
(function($) {

$.component.Container.extend('Repeater', 'content', {
    _init: function()
    {
        var props = arguments[0];
        // Take my children and give them to a "clone" template that will repeat once for each object
        var children = props.children || [];
        delete props.children;

        this._delayUntilVisible = true;

        // Normal object setup
        this._super.apply(this, arguments);

        // Record keeping preparation
        this._map = [];

        this._idPrefix = this.id + '-';

        setUpProperties(this, children);
    },

    _initDom: function()
    {
        this._super.apply(this, arguments);
        if ($.isBlank(this._properties.height) && (this._isDirty && !this.$dom.hasClass('serverRendered')))
        {
            this._tempHeight = 500;
            this.$dom.css('min-height', this._tempHeight);
        }
        // hook up the real container when rendered by the server
        if ($.isBlank(this._realContainer) && this._properties.container && $.isBlank(this.first))
        { this._setUpRealContainer(); }
    },

    _dataReady: function()
    {
        // We just need the dc loaded to hook up events; don't actually do anything
    },

    design: function(designing)
    {
        var cObj = this;
        // If we were designing, update from the edited config
        if (this._designing && !designing)
        {
            this._cloneProperties.children = this._readChildren();
            // Need to clean out a contextId when there is also a context since cId takes precedence,
            // but is probably wrong for the design item
            var cleanCId = function(item)
            {
                _.each(item.children, function(c)
                {
                    if (!$.isBlank(c.context) && !$.isBlank(c.contextId))
                    { delete c.contextId; }
                    cleanCId(c);
                });
            };
            cleanCId(this._cloneProperties);
        }

        this._super.apply(this, arguments);
        if ($.subKeyDefined(this, '_delayUntilVisible'))
        { this._delayUntilVisible = !this._designing; }
        this._render();
    },

    fetchChildren: function(start, count, callback)
    {
        var cObj = this;
        cObj.startLoading();
        cObj._loadRows(start, count, function(renderedRows)
        {
            renderedRows = _.select(renderedRows, function(r) { return $.isPlainObject(r); });
            callback(renderedRows);
            cObj.finishLoading();
        });
    },

    _clearDataContext: function()
    {
        var cObj = this;
        // Unbind anything old
        _.each($.makeArray(cObj._dataContext), function(dc)
        {
            if (dc instanceof DataContext)
            { dc.unbind(null, null, cObj); }
            if (!$.isBlank(dc.dataset))
            { dc.dataset.unbind(null, null, cObj); }
        });
        cObj._super.apply(cObj, arguments);
    },

    _addDataContext: function(dc)
    {
        var cObj = this;
        this._super.apply(this, arguments);

        if ($.subKeyDefined(dc, 'dataset') && this._properties.repeaterType != 'column')
        {
            dc.dataset.bind('query_change', function()
            {
                cObj._childrenDirty = true;
                cObj._isDirty = true;
                if (!$.isBlank(cObj.$dom))
                { cObj.$dom.removeClass('serverRendered'); }
                cObj._render();
            }, this);
            cObj.$dom.attr('aria-live', 'polite');
        }

        if (dc.type == 'datasetList')
        {
            dc.bind('data_change', function()
            {
                cObj._childrenDirty = true;
                cObj._isDirty = true;
                if (!$.isBlank(cObj.$dom))
                { cObj.$dom.removeClass('serverRendered'); }
                cObj._render();
            }, this);
            cObj.$dom.attr('aria-live', 'polite');
        }
    },

    _setUpRealContainer: function()
    {
        this._realContainer = this.add(this._properties.container || {type: 'Container'});
        delete this._containerDirty;
    },

    _loadRows: function(start, count, callback)
    {
        var cObj = this;
        var renderedRows = [];
        if ($.isBlank(cObj._dataContext))
        {
            callback();
        }
        else if (_.isArray(cObj._dataContext.value))
        {
            _.each(cObj._dataContext.value.slice(start, start + count), function(di, i)
            {
                if (!_.isObject(di))
                { di = { value: di }; }
                renderedRows.push(cObj._createRow(di, i + start, di));
            });
            callback(renderedRows);
        }
        else if (_.isArray(cObj._dataContext.list))
        {
            _.each(cObj._dataContext.list.slice(start, start + count), function(di, i)
            {
                if (!_.isObject(di))
                { di = { value: di }; }
                renderedRows.push(cObj._createRow(di, i + start, di));
            });
            callback(renderedRows);
        }
        else if (_.isArray(cObj._dataContext.goalList))
        {
            var goalItems = cObj._dataContext.goalList.slice(start, start + count);
            if (!$.isBlank(cObj._properties.groupBy))
            { renderGroupItems(cObj, goalItems, callback); }
            else
            {
                _.each(goalItems, function(di, i)
                {
                    if (!_.isObject(di))
                    { di = { value: di }; }
                    renderedRows.push(cObj._createRow(di, i + start, di));
                });
                callback(renderedRows);
            }
        }
        else if (cObj._dataContext.type == 'goalDashboard')
        {
            var catItems = cObj._dataContext.dashboard.categories.slice(start, start + count);
            if (!$.isBlank(cObj._properties.groupBy))
            { renderGroupItems(cObj, catItems, callback); }
            else
            {
                _.each(catItems, function(di, i)
                {
                    if (!_.isObject(di))
                    { di = { value: di }; }
                    renderedRows.push(cObj._createRow(di, i + start, di));
                });
                callback(renderedRows);
            }
        }
        else if (_.isArray(cObj._dataContext.categoryList))
        {
            var catItems = cObj._dataContext.categoryList.slice(start, start + count);
            if (!$.isBlank(cObj._properties.groupBy))
            { renderGroupItems(cObj, catItems, callback); }
            else
            {
                _.each(catItems, function(di, i)
                {
                    if (!_.isObject(di))
                    { di = { value: di }; }
                    renderedRows.push(cObj._createRow(di, i + start, di));
                });
                callback(renderedRows);
            }
        }
        else if (view = (cObj._dataContext || {}).dataset)
        {
            if (cObj._properties.repeaterType == 'column')
            {
                var exF = cObj._stringSubstitute(cObj._properties.excludeFilter);
                var incF = cObj._stringSubstitute(cObj._properties.includeFilter);
                _.each(_.select(cObj._dataContext.dataset.visibleColumns, function(c)
                {
                    if (_.all(exF, function(v, k)
                        { return !_.include($.makeArray(v), $.deepGetStringField(c, k)); }) &&
                        ($.isBlank(cObj._properties.includeFilter) ? true :
                                   _.any(incF, function(v, k)
                                       { return _.include($.makeArray(v), $.deepGetStringField(c, k)); })))
                    { return true; }
                    return false;
                }).slice(start, start + count), function(c, i)
                    { renderedRows.push(cObj._createRow(cObj._dataContext, i + start, {column: c})); });
                callback(renderedRows);
            }
            else
            {
                view.getRows(start, count, function(rows)
                {
                    rows = _.map(rows, function(r) { return view.rowToSODA2(r); });
                    if (!$.isBlank(cObj._properties.groupBy))
                    { renderGroupItems(cObj, rows, callback); }
                    else
                    {
                        callback(_.map(rows, function(r)
                            { return cObj._createRow(r, r[':index'], r); }));
                    }
                });
            }
        }
        else if (cObj._dataContext.type == 'datasetList')
        {
            cObj._dataContext.getItems(start, count, function(items)
            {
                callback(_.map(items, function(ds, i)
                        { return cObj._createRow(ds, i + start, $.extend({}, ds)); }));
            });
        }
        // Nothing to repeat on for row or column
    },

    _totalItems: function(callback)
    {
        var cObj = this;
        if ($.isBlank(cObj._dataContext))
        { callback(0); }
        else if (_.isArray(cObj._dataContext.value))
        { callback(cObj._dataContext.value.length); }
        else if (_.isArray(cObj._dataContext.list))
        { callback(cObj._dataContext.list.length); }
        else if (cObj._dataContext.type == 'goalDashboard')
        { callback(cObj._dataContext.dashboard.categories.length); }
        else if (_.isArray(cObj._dataContext.goalList))
        { callback(cObj._dataContext.goalList.length); }
        else if (_.isArray(cObj._dataContext.categoryList))
        { callback(cObj._dataContext.categoryList.length); }
        else if (view = (cObj._dataContext || {}).dataset)
        {
            if (cObj._properties.repeaterType == 'column')
            {
                var exF = cObj._stringSubstitute(cObj._properties.excludeFilter);
                var incF = cObj._stringSubstitute(cObj._properties.includeFilter);
                callback(_.select(cObj._dataContext.dataset.visibleColumns, function(c)
                {
                    if (_.all(exF, function(v, k)
                        { return !_.include($.makeArray(v), $.deepGetStringField(c, k)); }) &&
                        ($.isBlank(cObj._properties.includeFilter) ? true :
                                   _.any(incF, function(v, k)
                                       { return _.include($.makeArray(v), $.deepGetStringField(c, k)); })))
                    { return true; }
                    return false;
                }).length);
            }
            else
            {
                view.getTotalRows(function()
                { callback(view.totalRows()); });
            }
        }
        else if (cObj._dataContext.type == 'datasetList')
        { callback(cObj._dataContext.count); }
        else
        { callback(0); }
    },

    _refresh: function()
    {
        var cObj = this;
        if ($.isBlank(cObj._realContainer) || cObj._designing)
        {
            while (cObj.first)
            { cObj.first.destroy(); }
            _.each(cObj._funcChildren, function(fc) { fc.destroy(); });
            cObj._funcChildren = [];
            cObj._map = [];
            delete cObj._realContainer;
        }
        else if (!$.isBlank(cObj._realContainer))
        {
            if (cObj._containerDirty)
            {
                cObj._realContainer.destroy();
                delete cObj._realContainer;
                cObj._map = [];
            }
            else if (cObj._childrenDirty)
            {
                cObj._realContainer.empty();
                cObj._map = [];
            }
        }

        var view;
        if (cObj._designing)
        {
            // Render actual children as direct descendants
            cObj.add(cObj._cloneProperties.children);
            if (!$.isBlank(cObj._realContainer)) { cObj._realContainer.design(true); }
            cObj.each(function(c)
            { c.design(true); });

            if (!$.isBlank(cObj._realContainer)) { cObj._realContainer._designSubsidiary = true; }
            var addDesignSub = function(c)
            {
                if (_.isFunction(c.each))
                {
                    c.each(function(cc)
                    {
                        cc._designSubsidiary = true;
                        addDesignSub(cc);
                    });
                }
            };
            addDesignSub(cObj);
        }
        else
        {
            if ($.isBlank(cObj._realContainer))
            { cObj._setUpRealContainer(); }

            cObj._totalItems(function(totalCount)
            {
                var rowCount = cObj._stringSubstitute(cObj._properties.rowCount);
                var pageCount = rowCount || 100;
                var start = (cObj._stringSubstitute(cObj._properties.rowPage || 1) - 1) * pageCount;

                if (!$.isBlank(rowCount))
                { totalCount = Math.min(totalCount, rowCount); }

                if ($.subKeyDefined(cObj, '_realContainer.setCount'))
                { cObj._realContainer.setCount(totalCount); }

                var initCount;
                // Can't page with grouping
                if ($.subKeyDefined(cObj, '_realContainer.pageSize') && $.isBlank(cObj._properties.groupBy))
                { initCount = cObj._realContainer.pageSize(); }
                else
                { initCount = Math.min(totalCount, pageCount); }

                cObj.startLoading();
                cObj._loadRows(start, initCount, function(renderedRows)
                {
                    renderedRows = _.compact(renderedRows);
                    delete cObj._childrenDirty;
                    if (renderedRows.length < 1)
                    {
                        var clone = new $.component.Repeater.Clone($.extend(true, {},
                                cObj._noChildrenCloneProperties), cObj._componentSet);
                        // Insert the clone
                        cObj._realContainer.add(clone);
                    }
                    else
                    {
                        _.each(renderedRows, function(r)
                        {
                            if ($.isPlainObject(r))
                            { cObj._realContainer.add(r.row, r.index); }
                        });
                    }

                    cObj.finishLoading();
                    _.defer(function()
                    {
                        if (!$.isBlank(cObj._realContainer)) { cObj._realContainer._render(); }
                        $.component.sizeRenderRefresh();
                    });
                });
            });
        }
    },

    _render: function()
    {
        if (!this._super.apply(this, arguments)) { return false; }
        var cObj = this;

        var doRender = function()
        {
            if (!$.isBlank(cObj._tempHeight))
            {
                cObj.$dom.css('min-height', '');
                // Hack: This triggers a re-show of items
                $(window).scroll();
            }
            cObj._refresh();
        }
        if (!cObj._updateDataSource(cObj._properties, doRender))
        { doRender(); }

        return true;
    },

    _createRow: function(row, index, entity)
    {
        entity = entity || {};

        if ($.isBlank(entity._repeaterIndex)) { entity._repeaterIndex = index; }
        if ($.isBlank(entity._repeaterDisplayIndex)) { entity._repeaterDisplayIndex = index + 1; }
        if ($.isBlank(entity._evenOdd)) { entity._evenOdd = (index % 2) == 0 ? 'even' : 'odd'; }
        // Add ID prefix so repeated components will not clash
        var prefix = this._idPrefix + index + '-';
        function createTemplate(properties)
        {
            properties = _.clone(properties);
            properties.parentPrefix = prefix;
            properties.htmlClass = $.makeArray(properties.htmlClass);
            properties.htmlClass.push('id-' + properties.id);
            properties.id = prefix + (properties.id || $.component.allocateId());
            properties.entity = entity;
            var children = properties.children;
            if (properties.type != 'Repeater' && children)
            {
                if (!$.isArray(children))
                { children = [ children ]; }
                children = properties.children = children.slice();
                for (var i = 0; i < children.length; i++)
                { children[i] = createTemplate(children[i]); }
            }
            return properties;
        }
        var cloneProperties = createTemplate($.extend({},
                    this._stringSubstitute(this._properties.childProperties, entity),
                    this._cloneProperties));

        // Remove any existing row
        if (this._map[index])
        {
            if (!this._childrenDirty && this._map[index].id == cloneProperties.id)
            { return true; }
            this._map[index].remove();
            this._map[index] = undefined;
        }

        cloneProperties.childContextId = (row.id || '').toString();

        // Create clone
        var clone = this._map[index] = new $.component.Repeater.Clone(cloneProperties,
                this._componentSet);

        // Terrible hack; but core server doesn't support regexes in queries,
        // so this is the easiest way to skip some rows. This also doesn't
        // handle the fact that the number of rendered rows doesn't match up
        // directly with the position & length anymore; so it will probably not
        // behave properly if more than a page of data is present
        if (!$.isBlank(this._properties.valueRegex))
        {
            var r = new RegExp(this._stringSubstitute(this._properties.valueRegex.regex));
            var v = clone._stringSubstitute(this._properties.valueRegex.value);
            var result = r.test(v);
            if (this._properties.valueRegex.invert) { result = !result; }
            if (!result)
            {
                clone.destroy();
                return null;
            }
        }

        return { row: clone, index: index };
    },

    _readChildren: function()
    { return this._designing ? this._super() : this._cloneProperties.children; },

    _propWrite: function(properties)
    {
        var cObj = this;
        var origCloneProps = cObj._cloneProperties;
        // Make a copy so we don't screw up the original hash
        properties = $.extend(true, {}, properties);
        var origProps = $.extend(true, {}, cObj._properties);
        var children = properties.children;
        delete properties.children;

        cObj._super(properties);

        if (!$.isBlank(children))
        { setUpProperties(this, children); }

        cObj._containerDirty = !_.isEqual(cObj._properties.container, origProps.container);
        cObj._childrenDirty = !_.isEqual(cObj._cloneProperties, origCloneProps) ||
            _.any(['context', 'contextId', 'parentPrefix', 'groupBy',
                    'repeaterType', 'excludeFilter', 'includeFilter', 'valueRegex',
                    'childProperties', 'childHtmlClass', 'childStyles'], function(p)
            { return !_.isEqual(cObj._properties[p], origProps[p]); });

        if (cObj._containerDirty || cObj._childrenDirty)
        { cObj._render(); }
    }
});

$.component.Repeater.Clone = $.component.Container.extend({
    // No special behavior for clones at the moment
    _persist: false
});

var setUpProperties = function(cObj, children)
{
    cObj._cloneProperties = {
        id: 'clone',
        children: children,
        // TODO: Get rid of these out of templates so we can remove them
        htmlClass: cObj._properties.childHtmlClass,
        styles: cObj._properties.childStyles
    };

    cObj._noChildrenCloneProperties = {
        id: cObj.id + '_noChildren-clone',
        children:  cObj._properties.noResultsChildren
    };

    // Ensure that all descendants have an ID.  This ID is prefixed during object rendering.
    function allocateIds(children)
    {
        for (var i = 0; i < children.length; i++)
        {
            var child = children[i];
            if (child.id == undefined)
            { child.id = $.component.allocateId(); }
            if (child.children)
            { allocateIds(child.children); }
        }
    }
    allocateIds(cObj._cloneProperties);
    allocateIds(cObj._noChildrenCloneProperties);
    updateContainerPrefix(cObj);
};

var renderGroupItems = function(cObj, items, callback)
{
    var groupConfig = cObj._properties.groupBy;
    var groupIndex = {};
    var groups = [];
    _.each(items, function(item)
    {
        var group = cObj._stringSubstitute(groupConfig.value, item);
        if ($.isBlank(group)) { return; }
        var addGroupItem = function(g)
        {
            if ($.isBlank(groupIndex[g]))
            {
                groups.push(g);
                groupIndex[g] = [item];
            }
            else
            { groupIndex[g].push(item); }
        };

        if (!$.isBlank(groupConfig.splitOn))
        {
            _.each(group.split(groupConfig.splitOn), function(g)
            { addGroupItem(g); });
        }
        else { addGroupItem(group); }
    });

    if (groupConfig.sortAlpha)
    { groups = groups.sort(); }
    callback(_.map(groups, function(g, i)
    { return cObj._createRow({id: g}, i, {_groupValue: g, _groupItems: groupIndex[g]}); }));
};

var updateContainerPrefix = function(cObj)
{
    var parPrefix = cObj._properties.parentPrefix || '';
    if ($.subKeyDefined(cObj, '_properties.container.id') &&
            !cObj._properties.container.id.startsWith(parPrefix))
    {
        // Use the parent prefix so that if this is a top-level Repeater,
        // the realContainer id is unchanged
        cObj._properties.container.id = (cObj._properties.parentPrefix || '') +
            cObj._properties.container.id;
    }
};

})(jQuery);

File: lib/canvas2/widgets/repeater.rb
module Canvas2
  class Repeater < Container
    def initialize(props, parent = nil, resolver_context = nil)
      # Can't modify original properties (except to add an ID, if needed); so
      # we render with a copy
      @orig_props = props
      dup_props = props.dup
      dup_props.delete('children')
      super(dup_props, parent, resolver_context)
      props['id'] = @properties['id']

      allocate_ids(props['children'])
      @clone_props = {
        'id' => 'clone',
        'children' => props['children'],
        'htmlClass' => props['childHtmlClass'],
        'styles' => props['childStyles'],
        'type' => 'Container'
      }
    end

    def render_contents
      t = ''
      fully_rendered = true
      all_c = []
      if !context.blank?
        if context.is_a?(Array)
          context.each_with_index do |item, i|
            item = { value: item } if !item.is_a?(Hash)
            all_c << add_row(item, i, item.clone)
          end
        elsif context[:type] == 'list'
          context[:list].each_with_index do |item, i|
            item = { value: item } if !item.is_a?(Hash)
            all_c << add_row(item, i, item.clone)
          end
        elsif context[:type] == 'goalList'
          if !@properties['groupBy'].blank?
            all_c = render_group_items(context[:goalList])
          else
            context[:goalList].each_with_index {|g, i| all_c << add_row(g, i, g.clone) }
          end

        elsif context[:type] == 'goalDashboard'
          if !@properties['groupBy'].blank?
            all_c = render_group_items(context[:dashboard]['categories'])
          else
            context[:dashboard]['categories'].each_with_index {|c, i| all_c << add_row(c, i, c.clone) }
          end

        elsif context[:type] == 'govstatCategoryList'
          if !@properties['groupBy'].blank?
            all_c = render_group_items(context[:categoryList])
          else
            context[:categoryList].each_with_index {|c, i| all_c << add_row(c, i, c.clone) }
          end

        elsif context[:type] == 'datasetList'
          context[:datasetList].each_with_index {|ds, i| all_c << add_row(ds, i, ds.clone) }

        elsif context[:type] == 'dataset'
          if @properties['repeaterType'] == 'column'
            ex_f = string_substitute(@properties['excludeFilter'])
            ex_f = [] if ex_f.blank?
            inc_f = string_substitute(@properties['includeFilter'])
            inc_f = [] if inc_f.blank?
            context[:dataset].visible_columns.each_with_index do |c, i|
              if ex_f.all? {|k, v| !(Array.try_convert(v) || [v]).include?(Util.deep_get(c, k))} &&
                (@properties['includeFilter'].blank? ||
                 inc_f.any? {|k, v| (Array.try_convert(v) || [v]).include?(Util.deep_get(c, k))})
                all_c << add_row(context, i, {column: c})
              end
            end

          else
            row_count = [1, string_substitute(@properties['rowCount'] || 100).to_i].max
            row_page = [1, string_substitute(@properties['rowPage'] || 1).to_i].max
            if Canvas2::Util.no_cache
              rows = context[:dataset].get_rows(row_count, row_page, {}, false, !Canvas2::Util.is_private)
            else
              rows = context[:dataset].get_cached_rows(row_count, row_page, {}, !Canvas2::Util.is_private)
            end
            rows = rows[:rows].map { |row| context[:dataset].row_to_SODA2(row) }

            if !@properties['groupBy'].blank?
              all_c = render_group_items(rows)
            else
              rows.each_with_index do |r, i|
                all_c << add_row(r, i, r)
              end
            end
          end
        end

        all_c.compact!
      end

      child_timings = []
      if all_c.length > 0 || !@properties['noResultsChildren'].nil?
        cont_config = @properties['container'] || {'type' => 'Container'}
        cont_config['id'] ||= Canvas2::Util.allocate_id
        orig_id = cont_config['id']
        cont_config['id'] = (@properties['parentPrefix'] || '') + cont_config['id']
        real_c = CanvasWidget.from_config(cont_config, self)
        @orig_props['container'] = cont_config
        @orig_props['container']['id'] = orig_id
        # Enable hack to prepend special item to Repeater
        real_c.children ||= []
        real_c.children.concat(all_c.length > 0 ? all_c :
           @properties['noResultsChildren'].nil? ? nil :
           [CanvasWidget.from_config({ type: 'Container', id: self.id + '_noChildren-clone',
             children: @properties['noResultsChildren'] }.with_indifferent_access, self)])
        r = real_c.render
        t += r[0]
        fully_rendered &&= r[1]
        child_timings.push(r[2])
      end

      [t, fully_rendered, child_timings]
    end

    def child_resolver
      parent_resolver = self.parent.child_resolver() if !self.parent.blank?
      parent_resolver = Util.base_resolver() if parent_resolver.blank?
      lambda do |name|
        if !context.blank?
          keyed_c = {}
          context.is_a?(Array) ?
            context.each { |dc| keyed_c[dc[:id]] = dc if dc.is_a?(Hash) } : (keyed_c[context[:id]] = context)
          v = Util.deep_get(keyed_c, name)
        end
        v = parent_resolver.call(name) if v.blank?
        v
      end
    end

    def child_context
      nil
    end

  protected
    def add_row(row, index, resolutions = {})
      resolutions['_repeaterIndex'] = index
      resolutions['_repeaterDisplayIndex'] = index + 1
      resolutions['_evenOdd'] = (index % 2) == 0 ? 'even' : 'odd'

      child_props = string_substitute(@properties['childProperties'], resolutions)
      copy = create_copy({}.deep_merge(@clone_props).deep_merge(child_props.is_a?(Hash) ? child_props : {}),
                         self.id + '-' + index.to_s + '-', resolutions)
      copy['childContextId'] = row[:id]
      c = CanvasWidget.from_config(copy, self, resolutions)
      if @properties.has_key?('valueRegex')
        r = Regexp.new(c.string_substitute(@properties['valueRegex']['regex']))
        v = c.string_substitute(@properties['valueRegex']['value'])
        result = r.match(v).blank?
        result = !result if @properties['valueRegex']['invert']
        return nil if result
      end
      c
    end

    def render_group_items(items)
      g_config = @properties['groupBy']
      g_index = Hash.new
      groups = []
      items.each do |item|
        g = string_substitute(g_config['value'], item)
        next if g.blank?
        g = g_config['splitOn'].blank? ? [g] : g.split(g_config['splitOn'])
        g.each do |gg|
          if g_index[gg].blank?
            groups << gg
            g_index[gg] = [item]
          else
            g_index[gg] << item
          end
        end
      end

      all_c = []
      groups = groups.sort if g_config['sortAlpha']
      groups.each_with_index do |g, i|
        all_c << add_row({id: g}, i, {_groupValue: g, _groupItems: g_index[g]})
      end
      all_c
    end

    def allocate_ids(components)
      (components || []).each do |c|
        c['id'] = Util.allocate_id if c['id'].blank?
        allocate_ids(c['children'])
      end
    end

    def create_copy(component, id_prefix, resolutions)
      new_c = component.clone
      new_c['htmlClass'] = new_c['htmlClass'].is_a?(Array) ? new_c['htmlClass'].clone :
        [new_c['htmlClass']].compact
      new_c['htmlClass'] << 'id-' + new_c['id']
      new_c['parentPrefix'] = id_prefix
      new_c['id'] = id_prefix + new_c['id']
      new_c['entity'] = resolutions
      if new_c['type'] != 'Repeater' && new_c['children'].is_a?(Array)
        new_c['children'] = new_c['children'].map {|c| create_copy(c, id_prefix, resolutions)}
      end
      new_c
    end
  end
end
